tests/interlocked_win32_ut/interlocked_win32_ut.c

```diff
@@ -0,0 +1,708 @@
+        .SetReturn(INT32_MAX);
+
+    ///act
+    int32_t return_val = interlocked_compare_exchange(&destination, exchange, comperand);
+
+    ///assert
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+    ASSERT_ARE_EQUAL(int32_t, INT32_MAX, return_val);
+}
+
+/*Tests_SRS_INTERLOCKED_WIN32_43_013: [interlocked_compare_exchange_128 shall call InterlockedCompareExchange128 from windows.h.]*/
+/*Tests_SRS_INTERLOCKED_WIN32_43_014 : [interlocked_compare_exchange_128 shall return true if* comperand_result equals the original value of* destination.]*/
+TEST_FUNCTION(interlocked_compare_exchange_128_calls_InterlockedCompareExchange128_true_case)
+{
+    ///arrange
+    volatile int64_t* destination = (volatile int64_t*)_aligned_malloc(2 * sizeof(int64_t), 16);
+    int64_t* comperand_result = (int64_t*)_aligned_malloc(2 * sizeof(int64_t), 16);
+
+    destination[0] = INT64_MAX;
+    destination[1] = INT64_MIN;
+    comperand_result[0] = INT64_MAX;
+    comperand_result[1] = INT64_MIN;
+
+    int64_t exchange_high = 2;
+    int64_t exchange_low = 3;
+
+    STRICT_EXPECTED_CALL(mock_InterlockedCompareExchange128((volatile LONG64*)destination, (LONG64)exchange_high, (LONG64)exchange_low, (LONG64*)comperand_result))
+        .SetReturn(1);
+    ///act
+    bool return_val = interlocked_compare_exchange_128(destination, exchange_high, exchange_low, comperand_result);
+
+    ///assert
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+    ASSERT_IS_TRUE(return_val);
+
+    ///cleanup
+    _aligned_free((void*)destination);
+    _aligned_free(comperand_result);
+}
+
+/*Tests_SRS_INTERLOCKED_WIN32_43_013: [interlocked_compare_exchange_128 shall call InterlockedCompareExchange128 from windows.h.]*/
+/*Tests_SRS_INTERLOCKED_WIN32_43_063 : [interlocked_compare_exchange_128 shall return false if* comperand_result does not equal the original value of* destination.]*/
+TEST_FUNCTION(interlocked_compare_exchange_128_calls_InterlockedCompareExchange128_false_case)
+{
+    ///arrange
+    volatile int64_t* destination = (volatile int64_t*)_aligned_malloc(2 * sizeof(int64_t), 16);
+    int64_t* comperand_result = (int64_t*)_aligned_malloc(2 * sizeof(int64_t), 16);
+
+    destination[0] = INT64_MAX;
+    destination[1] = INT64_MIN;
+    comperand_result[0] = INT64_MAX;
+    comperand_result[1] = INT64_MIN;
+
+    int64_t exchange_high = 2;
+    int64_t exchange_low = 3;
+
+    STRICT_EXPECTED_CALL(mock_InterlockedCompareExchange128((volatile LONG64*)destination, (LONG64)exchange_high, (LONG64)exchange_low, (LONG64*)comperand_result))
+        .SetReturn(0);
+    ///act
+    bool return_val = interlocked_compare_exchange_128(destination, exchange_high, exchange_low, comperand_result);
+
+    ///assert
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
```

Comment:
>ASSERT_ARE_EQUAL [](start = 4, length = 16)

ASSERT seems to be lacking comperand_result[0], comperand_result[1]

Reply:
I test this in the integration tests.

---
In reply to: [427744374](https://github.com/Azure/azure-c-util/pull/43#discussion_r427744374) [](ancestors = 427744374)